src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -0,0 +1,298 @@
+      .clone()
+      .handler(new ChannelInitializer[SocketChannel]() {
+        def initChannel(ch: SocketChannel): Unit = {
+          val pipeline = ch.pipeline()
+          val sslHandler = sslClientCtx.newHandler(ch.alloc())
+          pipeline
+            .addLast("ssl", sslHandler) //This needs to first
+            .addLast("frameEncoder", new LengthFieldPrepender(4))
+            .addLast("frameDecoder", new LengthFieldBasedFrameDecoder(Int.MaxValue, 0, 4, 0, 4))
+            .addLast(new ByteArrayEncoder())
+            .addLast(new ChannelInboundHandlerAdapter() {
+              override def channelInactive(ctx: ChannelHandlerContext): Unit = {
+                deactivation.success(())
+              }
+
+              override def userEventTriggered(ctx: ChannelHandlerContext, evt: Any): Unit = {
+                evt match {
+                  case e: SslHandshakeCompletionEvent =>
+                    if (e.isSuccess) activation.success(ctx) else activation.failure(e.cause())
+                    log.debug(
+                      s"Ssl Handshake client channel from ${ctx.channel().localAddress()} " +
+                        s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id} and ssl status ${e.isSuccess}"
+                    )
+                  case _ =>
+                    log.debug(
+                      s"User Event client channel from ${ctx.channel().localAddress()} " +
+                        s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id}"
+                    )
+                }
+                super.userEventTriggered(ctx, evt)
+              }
+            })
+            .addLast(new MessageNotifier[M](messageSubject))
+        }
+      })
+
+    def initialize: Task[ClientChannelImpl[M]] = {
+      toTask(bootstrap.connect(inetSocketAddress)).map(_ => this)
+    }
+
+    override def sendMessage(message: M): Task[Unit] = {
+
+      Task
+        .fromFuture(activationF)
+        .map(ctx => {
+          log.debug(
+            s"Processing outbound message from local address ${ctx.channel().localAddress()} " +
+              s"to remote address ${ctx.channel().remoteAddress()} via channel id ${ctx.channel().id()}"
+          )
+
+          ctx.writeAndFlush(Unpooled.wrappedBuffer(codec.encode(message)))
+        })
+        .map(_ => ())
+    }
+
+    override def in: Observable[M] = messageSubject
+
+    override def close(): Task[Unit] = {
+      messageSubject.onComplete()
+      Task
+        .fromFuture(activationF)
+        .flatMap(ctx => toTask(ctx.close()))
+        .flatMap(_ => Task.fromFuture(deactivationF))
+    }
+  }
+
+  private val channelSubject = PublishSubject[Channel[InetMultiAddress, M]]()
+
+  private val workerGroup = new NioEventLoopGroup()
+
+  private val clientBootstrap = new Bootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioSocketChannel])
+    .option[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
```

Comment:
These values look like good candidates to be configurable from scalanet users.

Reply:
Sure, but we will only drive this stuff from use cases. If somebody needs the option, we add it. Otherwise we hard code it.